mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp

```diff
@@ -1560,11 +1574,32 @@ vectorizeAsTensorUnpackOp(RewriterBase &rewriter, tensor::UnPackOp unpackOp,
 
   ArrayRef<int64_t> innerDimPos = unpackOp.getInnerDimsPos();
   ArrayRef<int64_t> innerTiles = unpackOp.getStaticInnerTiles();
+  ArrayRef<int64_t> sourceShape = unpackTensorType.getShape();
+  bool useInBoundsInsteadOfMasking = false;
+  ArrayRef<int64_t> outerDimsPerm = unpackOp.getOuterDimsPerm();
+
+  auto destSize = unpackOp.getDestRank();
+
+  // initVectorShape is the shape of the vector that will be read from the
+  // source tensor. It is set like this: Let's say the sourceShape is 'M' and
+  // the vectorSize (VS) array is size 'N' where N <= M. Thus:
+  // - initVectorShape = sourceShape.take_front(N)
+  // - if outer_dims_perms is present: do that permutation on initVectorShape.
+  // - Multiply all the locations pointed by innerDimPos by the innerTileSize
+  //  attribute value.
+  SmallVector<int64_t> initVectorShape(sourceShape.take_front(destSize));
+  if (inputVectorSizes.empty()) {
+    if (!outerDimsPerm.empty())
+      applyPermutationToVector(initVectorShape, outerDimsPerm);
+    for (auto [i, pos] : llvm::enumerate(innerDimPos))
+      initVectorShape[pos] *= innerTiles[i];
+
+    inputVectorSizes = initVectorShape;
+    useInBoundsInsteadOfMasking = true;
+  }
```

Comment:
Is yet another variable, `initVectorShape`, really needed? Also, see my comment below.

```suggestion
  if (inputVectorSizes.empty()) {
    inputVectorSizes.resize(sourceShape.take_front(destSize))
    if (!outerDimsPerm.empty())
      applyPermutationToVector(inputVectorSizes, outerDimsPerm);
    for (auto [i, pos] : llvm::enumerate(innerDimPos))
      inputVectorSizes[pos] *= innerTiles[i];

    useInBoundsInsteadOfMasking = true;
  }
```

Reply:
We won't be able to modify/resize the `inputVectorShape` var since it's an `ArrayRef`, but we can point it to another variable. In that sense, it's needed.

Reply:
I think we can't do the trick because it is `ArrayRef` type. We can only apply permutation on `SmallVector`.

Reply:
I've tried to simplify it so we don't have the `else` block. Also, we can't use `vectorSizes = inputVectorSizes` SmallVectors can't point to ArrayRefs but the other way around is possible.